src/librustdoc/clean/types.rs

```diff
@@ -91,6 +92,56 @@ impl ExternalCrate {
         tcx.crate_name(self.crate_num)
     }
 
+    crate fn src_root(&self, tcx: TyCtxt<'_>) -> PathBuf {
+        match self.src(tcx) {
+            FileName::Real(ref p) => match p.local_path().parent() {
+                Some(p) => p.to_path_buf(),
+                None => PathBuf::new(),
+            },
+            _ => PathBuf::new(),
+        }
+    }
+
+    /// Attempts to find where an external crate is located, given that we're
+    /// rendering in to the specified source destination.
+    crate fn location(
+        &self,
+        extern_url: Option<&str>,
+        dst: &std::path::Path,
+        tcx: TyCtxt<'_>,
+    ) -> ExternalLocation {
+        use ExternalLocation::*;
+        // See if there's documentation generated into the local directory
+        let local_location = dst.join(&*self.name(tcx).as_str());
+        if local_location.is_dir() {
+            return Local;
+        }
+
+        if let Some(url) = extern_url {
+            let mut url = url.to_string();
+            if !url.ends_with('/') {
+                url.push('/');
+            }
+            return Remote(url);
+        }
+
+        // Failing that, see if there's an attribute specifying where to find this
+        // external crate
+        self.attrs
+            .lists(sym::doc)
+            .filter(|a| a.has_name(sym::html_root_url))
+            .filter_map(|a| a.value_str())
+            .map(|url| {
+                let mut url = url.to_string();
+                if !url.ends_with('/') {
+                    url.push('/')
+                }
+                Remote(url)
```

Comment:
Feel free to ignore this in case this is out of scope or anything but I can see that this is done twice. Why not refactor this into a nice `to_remote` now that you are at it?
```rust
fn to_remote(url: &str) -> ExternalLocation {
    let mut url = url.to_string();
    if !url.ends_with('/') {
        url.push('/');
    }
    Remote(url)
}
```
and use it like this here: `.map(|url| to_remote(url))` and up there: `return to_remote(url);`.

Reply:
I would make this a closure instead so it's not as verbose, but that seems useful :)

Reply:
Actually I had to be a function so we can use a generic type `T: impl ToString` as argument. Because `url` is `&str` in the first call, but `rust_span::symbo::Symbol` in the second one.